### PR TITLE
[FEA] Generalized PDE residuals

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description
This PR adds utilities to compute PDE residuals on a variety of different model outputs and using a few different methods like autodiff, meshless finite difference, finite difference, spectral derivatives and least squares method. The idea is to enable addition of physics based residual calculations to the training loss as well as use these utilities to compute metrics during inference to access how well the predictions respect the underlying physics. 

Examples demonstrating these utilities will be added in a separate PR and cross-linked later. 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus-sym/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus-sym/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus-sym/issues) is linked to this pull request.

### CODE REVIEW

**Review by Alexey-Kamenev — APPROVED** [review]
LGTM!

**Review by RishikeshRanade — APPROVED** [review]
Looks good

File: modulus/sym/eq/mfd/grads.py [hunk 1]
```diff
@@ -378,12 +394,22 @@ def __init__(
...
+                    key.derivatives[0] == key.derivatives[1] == key.derivatives[2]
+                ), f"Derivatives keys should be the same"
+                eval_list.append(
+                    ThirdDerivO2(key.name, str(key.derivatives[0]), str(key))
+                )
 
         self._eval = torch.nn.ModuleList(eval_list)
 
 
 class ForthDeriv(DerivBase):
```
Comment (akshaysubr) [inline modulus/sym/eq/mfd/grads.py:410]: Forth -> Fourth

**Review by akshaysubr — APPROVED** [review]

File: modulus/sym/eq/fd/grads.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from typing import Union, List
+
+Tensor = torch.Tensor
+
+
+class FirstDerivO2(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (loliverhennigh) [reply modulus/sym/eq/fd/grads.py:23]: I have found the naming convention they use here fairly good, https://github.com/tumaer/JAXFLUIDS/blob/main/src/jaxfluids/stencils/derivative/deriv_center_2.py

File: modulus/sym/eq/fd/grads.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    """Module to compute first derivative with 2nd order accuracy"""
+
+    def __init__(self, dim: int, dx: Union[float, List[float]]):
+        super().__init__()
+        self.dim = dim
+        if isinstance(dx, float):
+            dx = [dx for _ in range(dim)]
+        self.dx = dx
+
+        self.register_buffer("ddx1D", torch.Tensor([-1.0 / 2.0, 0.0, 1.0 / 2.0]))
```
[the remark this answers is not in the document]
Comment (loliverhennigh) [reply modulus/sym/eq/fd/grads.py:33]: If we get to this point I would highly recommend writing warp kernels for these. We can have the same factory functionality and it will give us ~4x speed increase. Not needed for this PR though.

File: modulus/sym/eq/fd/grads.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ktangsali) [reply modulus/sym/eq/fd/grads.py]: Thanks, fixed!

File: modulus/sym/eq/fd/grads.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktangsali) [reply modulus/sym/eq/fd/grads.py:33]: Thanks! yes I agree that higher orders are needed, but I wanted to prioritize the structure in this PR. I have made this issue to keep a track of this: https://github.com/NVIDIA/modulus-sym/issues/180

I am marking this resolved currently :)
